handbook/engineering/testing.md

```diff
@@ -2,11 +2,87 @@
 
 This file documents how we test code at Sourcegraph.
 
+## Philosophy
```

Comment:
I'm not confident on where to fit this in, but it's worth pointing out why tests are important beyond quality of our product (which feels just a tad overly customer-centric), and what level of quality assurance it gives us. Feel free to incorporate as much or as little as the following bits that resonate with you:

> "Program testing can be used to show the presence of bugs, but never to show their absence!" - Dijkstra

- Ergo, testing allows us to find bugs, that's why they're so useful, and that's why we want them!
- A massive benefit of a well-tested code base is that it provides confidence for refactors and identifies gaps if a refactor breaks things. Refactoring and tests go hand-in-hand.
- Testing is a form of software development "self-care": we don't just write tests to feel good about shipping a feature _now_. We also write tests for our future ourselves and other team members because tests are a knowledge base for ourselves, a coarse and growing specification of what the code should do.

Reply:
> Ergo, testing allows us to find bugs, that's why they're so useful, and that's why we want them!

I think this is covered by the first documented point "Our product works as intended when we ship it to customers." Not working as intended is the definition of a bug :)

> Testing is a form of software development "self-care": we don't just write tests to feel good about shipping a feature now. We also write tests for our future selves and other team members because tests are a knowledge base for ourselves, a coarse and growing specification of what the code should do.

I think this is covered by the second documented point "Our product continues to work and doesn't accidentally break as we make changes to our product over time."

> A massive benefit of a well-tested code base is that it provides confidence for refactors and identifies gaps if a refactor breaks things. Refactoring and tests go hand-in-hand.

I think this is also covered by the second point, but I did add an extra sentence to this effect in d2a90d97bdb7faa26285d779612c014de61e83f4. I also changed "our product" to "our product and code" so it would sound less product centric (although our code only exists to serve the product).